addons/mail/tests/common.py

```diff
@@ -120,3 +120,34 @@ def tearDownClass(cls):
         cls.env['ir.mail_server']._revert_method('build_email')
         cls.env['ir.mail_server']._revert_method('send_email')
         super(TestMail, cls).tearDownClass()
+
+    def gateway_reply_wrecord(self, template, record, extra_irt=True):
+        """ Simulate a reply through the mail gateway. Usage: giving a record,
+        find an email sent to him and use its message-ID to simulate a reply. """
+        email = self._find_sent_email_wrecord(record)
+
+        if extra_irt:
+            extra = 'In-Reply-To:\r\n\t%s\n' % email['message_id']
+        else:
+            disturbing_other_msg_id = '<123456.654321@another.host.com>'
+            extra = 'References:\r\n\t%s\n\r%s' % (email['message_id'], disturbing_other_msg_id)
+
+        return self.format_and_process(
+            template, email_from=email['email_to'][0], to=email['reply_to'],
+            subject='Re: %s' % email['subject'],
+            extra=extra,
+            msg_id='<123456.%s.%d@test.example.com>' % (record._name, record.id),
+            target_model=record._name,
+            target_field=record._rec_name,
+        )
+
+    def _find_sent_email_wrecord(self, record):
+        """ Helper to find in outgoing emails (see build_email) an email linked to
+        a given record. It has been introduced with a fix for mass mailing and is
+        not meant to be used widely, proper tools are available in later versions. """
+        for mail in self._mails:
+            if mail['object_id'] == '%d-%s' % (record.id, record._name):
+                break
+        else:
+            raise AssertionError('Sent email not found for record %s' % record)
+        return mail
```

Comment:
Wait. `mail` is outside the loop where it's defined and you can return it?
Learn every day.

Reply:
for-else construct, people hates that :D